src/Workspaces/Core/Portable/SymbolKey/SymbolKey.cs

```diff
@@ -75,114 +74,140 @@ namespace Microsoft.CodeAnalysis
+        /// from any other source is not supported.
+        /// </summary>
+        public SymbolKey(string data)
         {
-            _symbolKeyData = symbolKeyData ?? throw new ArgumentNullException();
+            _symbolKeyData = data ?? throw new ArgumentNullException();
         }
 
-        public static IEqualityComparer<SymbolKey> GetComparer(bool ignoreCase, bool ignoreAssemblyKeys)
+        /// <summary>
+        /// Constructs a new <see cref="SymbolKey"/> representing the provided <paramref name="symbol"/>.
+        /// </summary>
+        public SymbolKey(ISymbol symbol, CancellationToken cancellationToken = default)
+            : this(CreateString(symbol, cancellationToken))
         {
-            return SymbolKeyComparer.GetComparer(ignoreCase, ignoreAssemblyKeys);
         }
 
-        private static readonly Func<string, string> s_removeAssemblyKeys = (string data) =>
-        {
-            var reader = new RemoveAssemblySymbolKeysReader();
-            reader.Initialize(data);
-            return reader.RemoveAssemblySymbolKeys();
-        };
+        internal static SymbolKey Create(ISymbol symbol, CancellationToken cancellationToken = default)
+            => new SymbolKey(symbol, cancellationToken);
 
         /// <summary>
-        /// Tries to resolve the provided <paramref name="symbolKey"/> in the given 
-        /// <paramref name="compilation"/> to a matching symbol.  <paramref name="resolveLocations"/>
-        /// should only be given <see langword="true"/> if the symbol was produced from a compilation
-        /// that has the exact same source as the compilation we're resolving against.  Otherwise
-        /// the locations resolved may not actually be correct in the final compilation.
+        /// Returns an <see cref="IEqualityComparer{T}"/> that determines if two <see cref="SymbolKey"/>s
+        /// represent the same effective symbol.
         /// </summary>
-        public static SymbolKeyResolution Resolve(
+        /// <param name="ignoreCase">Whether or not casing should be considered when comparing keys. 
+        /// For example, with <c>ignoreCase=true</c> then <c>X.SomeClass</c> and <c>X.Someclass</c> would be 
+        /// considered the same effective symbol</param>
+        /// <param name="ignoreAssemblyKeys">Whether or not the originating assembly of referenced
+        /// symbols should be compared when determining if two symbols are effectively the same.
+        /// For example, with <c>ignoreAssemblyKeys=true</c> then an <c>X.SomeClass</c> from assembly 
+        /// <c>A</c> and <c>X.SomeClass</c> from assembly <c>B</c> will be considered the same
+        /// effective symbol.
+        /// </param>
+        internal static IEqualityComparer<SymbolKey> GetComparer(bool ignoreCase = false, bool ignoreAssemblyKeys = false)
+            => SymbolKeyComparer.GetComparer(ignoreCase, ignoreAssemblyKeys);
+
+        internal static SymbolKeyResolution ResolveString(
```

Comment:
Seems we are using term string and data interchangeably in the APIs, maybe consider keeping them consistent.

Reply:
sounds good.